migrate kotlin tests from java mockito to mockk

theexiile1305 opened the pull request:
This pull requests migrate kotlin tests from java mockito to mockk within the config folder.  

Closes gh-9785

eleftherias reviewed (changes requested):
Thanks @theexiile1305!
There are some failing tests, could you please take a look at them?

theexiile1305 commented:
@eleftherias Yea, of course. I take a look at them.

theexiile1305 commented:
@eleftherias Can you provide me some information, how to execute the complete tests within Intellij 2021.1.1 Ultimate Edition or on the command line? 
The command `./gradlew clean build integrationTest` produces the following errors on my local mac (Big Sur):
```
> Task :buildSrc:checkApolloVersions UP-TO-DATE
> Task :buildSrc:generateMainServiceApolloSources UP-TO-DATE
> Task :buildSrc:checkMainServiceApolloDuplicates UP-TO-DATE
> Task :buildSrc:generateTestServiceApolloSources NO-SOURCE
> Task :buildSrc:checkTestServiceApolloDuplicates UP-TO-DATE
> Task :buildSrc:compileJava UP-TO-DATE
> Task :buildSrc:compileGroovy UP-TO-DATE
> Task :buildSrc:pluginDescriptors UP-TO-DATE
> Task :buildSrc:processResources UP-TO-DATE
> Task :buildSrc:classes UP-TO-DATE
> Task :buildSrc:jar UP-TO-DATE
> Task :buildSrc:assemble UP-TO-DATE
> Task :buildSrc:pluginUnderTestMetadata UP-TO-DATE
> Task :buildSrc:compileTestJava UP-TO-DATE
> Task :buildSrc:compileTestGroovy UP-TO-DATE
> Task :buildSrc:processTestResources UP-TO-DATE
> Task :buildSrc:testClasses UP-TO-DATE
> Task :buildSrc:test
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/Users/mfuchs/.gradle/wrapper/dists/gradle-6.9-bin/2ecsmyp3bolyybemj56vfn4mt/gradle-6.9/lib/groovy-all-1.3-2.5.12.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
io.spring.gradle.convention.JavadocApiPluginITest > multimodule api FAILED
    org.spockframework.runtime.ConditionFailedWithExceptionError at JavadocApiPluginITest.groovy:44
        Caused by: java.io.FileNotFoundException at JavadocApiPluginITest.groovy:44
io.spring.gradle.convention.JacocoPluginITest > check with java plugin FAILED
    org.gradle.testkit.runner.UnexpectedBuildFailure at JacocoPluginITest.groovy:30
io.spring.gradle.convention.ShowcaseITest > build FAILED
    org.gradle.testkit.runner.UnexpectedBuildFailure at ShowcaseITest.groovy:32
63 tests completed, 3 failed, 3 skipped
> Task :buildSrc:test FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':buildSrc:test'.
> There were failing tests. See the report at: file:///Users/mfuchs/git/github/spring-security/buildSrc/build/reports/tests/test/index.html
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
BUILD FAILED in 53s
A build scan was not published as you have not authenticated with server 'ge.spring.io'.
```

What is my mistake?

eleftherias commented:
@theexiile1305 Is there any additional information in the report at `reports/tests/test/index.html`?

theexiile1305 commented:
@eleftherias I've resolved my issue from my  previous comment. The solution was the configuration of gradle and the project itself with java 11 instead of instead 14.

Now, I've fixed all test expect four in the classes `RememberMeDslTests` and `SessionManagementDslTests`.  Also I've added an TODO comment describing the current difference in the test result. Can you help me there please out or give me some hints?

theexiile1305 commented:
Now only the last three tests in `SessionManagementDslTests` fails. We need to setup a mock authentication, because I get a 403 instead of 302 http status. The depending tests are:
- `session management when session authentication error url then redirected to url`
- `session management when session authentication strategy then strategy used`
- `session management when session authentication failure handler then handler used`

eleftherias commented on config/src/test/kotlin/org/springframework/security/config/web/servlet/SessionManagementDslTests.kt:
```diff
@@ -29,31 +34,76 @@ import org.springframework.security.config.annotation.web.configuration.WebSecur
+
+    @Before
+    fun setUp() {
+        every { session.changeSessionId() } throws SessionAuthenticationException("any SessionAuthenticationException")
```
Let's keep the mocks in each test in order to avoid test pollution.

theexiile1305 replied:
I'll will change it in the next commit.

eleftherias commented on config/src/test/kotlin/org/springframework/security/config/web/servlet/SessionManagementDslTests.kt:
```diff
@@ -29,31 +34,76 @@ import org.springframework.security.config.annotation.web.configuration.WebSecur
+        every { session.getAttribute(FLASH_MAPS_SESSION_ATTR_NAME) } returns emptyList<Any>()
+        every { session.removeAttribute(FLASH_MAPS_SESSION_ATTR_NAME) } returns Unit
+        every {
+            session.getAttribute(HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY)
```
This line is what is causing the test `session management when session authentication error url then redirected to url` to fail. The `SPRING_SECURITY_CONTEXT_KEY` attribute must be null in order to reach the code block that causes the authentication error.
https://github.com/spring-projects/spring-security/blob/68f91edbb8a223ee8505893697d85cbe51890ea2/web/src/main/java/org/springframework/security/web/session/SessionManagementFilter.java#L91-L105
We may not be able to use Mockk in this case, because I believe it doesn't support returning null values.

theexiile1305 replied:
Yea, you are right. Mockk doesn't support `null` as a returning type. This makes really sense, because kotlin is designed for the null-saftey. I rollbacked the  testclass `SessionManagementDslTests`, so that Mockito is used.
Thank you very much for the help.

eleftherias reviewed (changes requested):
Thanks for the updates @theexiile1305!

I have left a comment inline about what the issue is in one of the tests.
I believe the rest of the tests have a similar issue.

We may have to keep Mockito for some mocks that have to return null.
Feel free to investigate a workaround using Mockk or continue using Mockito in those cases.

theexiile1305 commented:
@eleftherias Finally, I successfully have migrated mockito to mockk in the test class `SessionManagementDslTests.kt`. Now everything is ready for a review 😃

eleftherias commented on config/src/test/kotlin/org/springframework/security/config/web/server/ServerCsrfDslTests.kt:
```diff
@@ -161,70 +160,72 @@ class ServerCsrfDslTests {
-        `when`(MultipartFormDataNotEnabledConfig.TOKEN_REPOSITORY.generateToken(any()))
-                .thenReturn(Mono.just(this.token))
         this.spring.register(MultipartFormDataNotEnabledConfig::class.java).autowire()
+        val multipartFormDataNotEnabledConfig = this.spring.context.getBean(MultipartFormDataNotEnabledConfig::class.java)
+        every {
+            multipartFormDataNotEnabledConfig.serverCsrfTokenRepository.loadToken(any())
+        } returns Mono.just(this.token)
+        every {
+            multipartFormDataNotEnabledConfig.serverCsrfTokenRepository.generateToken(any())
+        } returns Mono.just(this.token)
```
In order to be consistent with our Java tests, we should continue using companion objects rather than getting the bean from the context. 
It seems with MockK we cannot create a mock in the companion object, but we can use the `mockkObject` to achieve the same behaviour.
```suggestion
        mockkObject(MultipartFormDataNotEnabledConfig.TOKEN_REPOSITORY)
        every {
            MultipartFormDataNotEnabledConfig.TOKEN_REPOSITORY.loadToken(any())
        } returns Mono.just(this.token)
        every {
            MultipartFormDataNotEnabledConfig.TOKEN_REPOSITORY.generateToken(any())
        } returns Mono.just(this.token)
```

eleftherias commented on config/src/test/kotlin/org/springframework/security/config/web/server/ServerCsrfDslTests.kt:
```diff
@@ -237,27 +238,29 @@ class ServerCsrfDslTests {
-            var TOKEN_REPOSITORY: ServerCsrfTokenRepository = mock(ServerCsrfTokenRepository::class.java)
-        }
+
+        val serverCsrfTokenRepository: ServerCsrfTokenRepository = mockk()
```
Since we can't use a mock or null here, we must create an actual instance.
```suggestion
       companion object {
            var TOKEN_REPOSITORY: ServerCsrfTokenRepository = WebSessionServerCsrfTokenRepository()
        }
```

eleftherias reviewed (changes requested):
Thanks @theexiile1305!
I have left some comments inline regarding companion objects.
Please keep the use of companion objects wherever possible.
I added an example of how we can make them work with Mockk.

eleftherias commented:
Great work on your first contribution @theexiile1305!
The updates to `ServerCsrfDslTests.kt` look good to me.

To answer your questions:
- The main idea is to keep consistency across the tests. I also find it more readable to mock the specific object that we need, instead of getting the configuration using `this.spring.context.getBean`. If you feel strongly that we shouldn't use companion objects, we can discuss changing it.
- Yes, when you are finished please squash all commits and include `Closes gh-9785` at the end of your [commit message](https://github.com/spring-projects/spring-security/blob/main/CONTRIBUTING.adoc#format-commit-messages).

theexiile1305 commented:
@eleftherias Thanks for that feedback and I understand the point to keep consistency across the tests. I agree with that 😄 

I updated all tests to keep the mocks in the companion objects. If all changes looks good to you, I will squash all commits into one including `Closes gh-9785`.

eleftherias commented on config/src/test/kotlin/org/springframework/security/config/web/servlet/RememberMeDslTests.kt:
```diff
@@ -312,24 +329,39 @@ internal class RememberMeDslTests {
+            RememberMeDefaultUserDetailsServiceConfig.PASSWORD_ENCODER.matches(any(), any())
+        } returns true
+        every {
+            RememberMeDefaultUserDetailsServiceConfig.PASSWORD_ENCODER.encode(any())
```
We can use a real password encoder `BCryptPasswordEncoder()` instead of a mock.

eleftherias commented on config/src/test/kotlin/org/springframework/security/config/web/servlet/Saml2DslTests.kt:
```diff
@@ -1,5 +1,5 @@
 /*
```
Please revert the changes to this file, since the only changes are import ordering.

eleftherias commented on config/src/test/kotlin/org/springframework/security/config/web/servlet/session/SessionFixationDslTests.kt:
```diff
@@ -26,10 +26,10 @@ import org.springframework.mock.web.MockHttpSession
 import org.springframework.security.config.annotation.web.builders.HttpSecurity
```
Please revert the changes to this file, since the only changes are import ordering.

eleftherias reviewed (changes requested):
Thanks @theexiile1305! I have left some comments inline.

theexiile1305 commented:
@eleftherias Sorry for the files, whereas the import orders were changed that was an accident by intellij.  
Also I've used the  real password encoder `BCryptPasswordEncoder()` inside the test. Saidly, there is no way to keep the verification that the success handler was called on the request to "/abc" and not on the request to "/login". 

I think the best way to review the last changes is to keep the latest three commits. If everything looks fine, I will squash them for sure.

eleftherias commented:
Thanks @theexiile1305! All the updates look good.
Please go ahead and squash the commits and I will merge this PR.

theexiile1305 commented:
Thanks @eleftherias for the review! 
Also I squashed all commits 😄

eleftherias reviewed (changes requested):
Awesome @theexiile1305!
One final comment about the commit message, could you update it to say:
```
Migrate Kotlin tests from java Mockito to Mockk

Closes gh-9785
```
This follows the case and spacing conventions that we typically use in commit messages.

theexiile1305 commented:
> Awesome @theexiile1305!
> One final comment about the commit message, could you update it to say:
> 
> ```
> Migrate Kotlin tests from java Mockito to Mockk
> 
> Closes gh-9785
> ```
> 
> This follows the case and spacing conventions that we typically use in commit messages.

Done 😃

eleftherias commented:
Thanks for the PR @theexiile1305! This is now merged into the main branch.
